Create HexColorField component

emircado opened the pull request:
Closes #974 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component). **Will not write documentation as advised**
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Do necessary tests done on new components
1. Should work fine across supported browsers, screen readers
2. Should increment and decrement with keyboard/mouse wheel
3. Should have accessibility features (aria) needed for a `spinbutton` role
4. Should have component-specific features working, such as custom step value
5. Should trigger `onChange` events only when necessary
6. Should accept only valid hex characters `[#0-9a-fA-F]`

## 🧢 Your Project:

B2BCMP

emircado commented on packages/@react-spectrum/color/src/HexColorField.ts:
```diff
@@ -0,0 +1,25 @@
+/**
+ * HexColorField allows users to input a color value in hex format.
+ */
+const _HexColorField = React.forwardRef(HexColorField);
```
TODO: Ask when it's appropriate to do `React.forwardRef()` for a component

dannify replied:
Pretty much every component 
👍

emircado commented on packages/@react-spectrum/color/src/index.ts:
```diff
@@ -0,0 +1,15 @@
+ * governing permissions and limitations under the License.
+ */
+
+/// <reference types="css-module-types" />
```
TO DO: Ask what this does

mischnic replied:
It's needed so that running `make build` doesn't print Typescript errors about unresolved CSS imports such as these:
https://github.com/adobe/react-spectrum/blob/2eddd0b274b9f2c418a40ad535aa0164757a76bb/packages/%40react-spectrum/button/src/Button.tsx#L19
If you don't import any CSS in e.g. `HexColorField.ts`, it can be removed.

emircado replied:
Thanks!

emircado commented on packages/@react-types/color/src/index.d.ts:
```diff
@@ -0,0 +1,41 @@
+  TextInputDOMProps,
+  Validation,
+} from '@react-types/shared';
+import {Color} from '@react-stately/color';
```
Doing this felt kinda weird. Is this fine?

I was imagining `Color` from `react-stately` would extend an interface we define here in `react-types`, and that interface is also what we'll refer to as `Color` here in `HexColorFieldProps`.

I did this because I saw something similar from `@react-types/sidenav`
```
import {ReusableView} from '@react-stately/virtualizer';
```

devongovett replied:
Yeah, we should make an interface for `Color` in `@react-types/color` and have the class in `@react-stately/color`  implement it.

emircado commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -10,13 +10,13 @@
+import {
+  ColorFormat,
+  ColorChannel,
+  Color as ColorType,
```
I wonder if it's okay to leave a `Color` interface in `@react-types/color` and `Color` class in `@react-stately/color`

devongovett replied:
Yep, that's correct!

emircado commented on packages/@react-stately/color/test/useHexColorFieldState.test.js:
```diff
@@ -0,0 +1,54 @@
+import {renderHook} from '@testing-library/react-hooks';
+import {useHexColorFieldState} from '../';
+
+describe('useHexColorFieldState tests', function () {
```
Add test for invalid value. Expect error to be thrown

emircado replied:
Resolved. Added test "should be in invalid state"

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,84 @@
+
+      const maxColor = typeof maxValue === 'string'? new Color(maxValue) : maxValue;
+      const maxColorString = maxColor.toString('hex').substring(1);
+      const maxColorNumber = parseInt(maxColorString, 16);
```
Might be worth writing a util for common code

emircado replied:
Resolved. Added utils in `Color.ts`

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,84 @@
+      const maxColorString = maxColor.toString('hex').substring(1);
+      const maxColorNumber = parseInt(maxColorString, 16);
+
+      const newValue = `#${Math.min(colorNumber + step, maxColorNumber).toString(16)}`;
```
I thought about using `clamp` from `useNumberFieldState` but `handleDecimalOperation` not being exported

emircado replied:
@snowystinger:
This is fine, as we don't have to worry about decimal operations here

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,142 @@
+    decrement,
+    decrementToMin,
+    commitInputValue,
+    validationState: !isValid ? 'invalid' : null,
```
Need to check this with the team. I copied this from `NumberField` but this kind of suggests the component will not be in a `valid` state (with the green check icon)

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,150 @@
+    setIsValid(isInputValueValid(value, maxValue, minValue));
+  };
+
+  const commitInputValue = () => {
```
I'm not quite sure how to write a test for this. I want to show how `colorValue` reverts back to its last valid value when it receives a new value through `setColorValue`. It's just that `setColorValue` strictly accepts `Color` objects and will always be valid afaik.

emircado replied:
Resolved. Added test `should revert to last valid value`

emircado commented on packages/@react-stately/color/test/useHexColorFieldState.test.js:
```diff
@@ -0,0 +1,169 @@
+    expect(result.current.colorValue).toBeUndefined();
+    expect(result.current.inputValue).toBe('');
+    expect(result.current.validationState).toBe('invalid');
+  });
```
Do we really want this to return an invalid state instead of throwing an error?

emircado replied:
@snowystinger 
Depends on what invalid input here is. Consider invalid color string as empty and clamp it to min

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,150 @@
+
+  return {
+    colorValue,
+    setColorValue,
```
I should write a test for this

emircado replied:
Resolved. Test added "should update colorValue"

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,150 @@
+    onChange
+  } = props;
+
+  const [colorInputValue, setColorInputValue] = useControlledState<ColorInput>(value, defaultValue || minValue, onChange);
```
NOTE TO SELF: Add test for controlled value

emircado replied:
Resolved. Added test "should have a controlled value"

emircado commented on packages/@react-stately/color/test/useHexColorFieldState.test.js:
```diff
@@ -0,0 +1,182 @@
+import {useHexColorFieldState} from '../';
+
+describe('useHexColorFieldState tests', function () {
+  it('should be at minimum if no value is provided', function () {
```
Need to confirm this with the team. `NumberField` starts at 0/min when no value is provided in the props

devongovett replied:
Would it make more sense for the field to be empty or set to the minimum by default?

emircado replied:
I'd prefer the input to be empty before you interact with it. Then once you focus on it the first time, it'll show the minimum value if there's no value/default set. I was hoping someone from our UX team could confirm what the behavior should be

emircado replied:
@snowystinger 
Should start empty unless a defaultValue or value is provided.

emircado commented on packages/@react-stately/color/test/useHexColorFieldState.test.js:
```diff
@@ -0,0 +1,218 @@
+      blue: 204,
+      alpha: 1
+    });
+    // expect(result.current.inputValue).toBe('#CCBBAA');
```
I hope I'm understanding the concept of controlled states correctly.

Here, `colorValue` is controlled so it doesn't change the `colorValue` state even after `setColorValue` is called.
Instead, it will rely on the component to be re-rendered for it to reflect the new color value.

I'm not sure how the other states like `inputValue` should respond. It currently still updates to the new color we provide in `setColorValue`. Maybe I should put it in a `useEffect` so `inputValue` changes only when `colorValue` changes.

emircado replied:
Resolved. Will revisit this question later if it still makes sense.

emircado commented on packages/@react-stately/color/docs/useHexColorFieldState.mdx:
```diff
@@ -0,0 +1,40 @@
+keywords: [color, hex, state]
+---
+
+# useHexColorFieldState
```
I assume all deliverables should be documented

emircado replied:
@snowystinger 
Delete doc things, we don't want it visible while in alpha

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,85 @@
+  });
+
+  const {
+    labelProps, 
```
`NumberField` that I used as a pattern for this component is also having issues with label not showing in storybook. Will have to revisit this

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,85 @@
+      'aria-label': props['aria-label'] || null,
+      'aria-labelledby': props['aria-labelledby'] || null,
+      id: inputId,
+      // placeholder: formatMessage('TO DO'),
```
Ask team for requirements around the default placeholder value

emircado replied:
@snowystinger 
leave it empty, but make sure users can add placeholder

emircado replied:
Resolved. Added a story for when it has a placeholder.

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,85 @@
+    labelProps, 
+    inputProps: inputFieldProps
+  } = useTextField(
+    mergeProps(focusProps, {
```
<img width="593" alt="Screen Shot 2020-09-03 at 3 05 18 AM" src="https://user-images.githubusercontent.com/5743833/92102509-8b398700-ed93-11ea-947a-8a189ec72a52.png">

I get this TS error/warning consistently where we pass `mergeProps(...)` in `useTextField`. It's because `useTextField` expects it to be `AriaTextFieldOptionsType` but instead it is `HTMLAttributes<...>`

emircado replied:
Will revisit this later. Removed this code and will see if it makes sense to add back later.

emircado replied:
@snowystinger 
go back to using mergeProps

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,154 @@
+    decrement,
+    decrementToMin,
+    commitInputValue,
+    validationState: !isValid ? 'invalid' : null
```
I remembered other form inputs don't necessarily decide their validationState on their own. I wonder if we should do the same here as well.

emircado replied:
Also, I copied this from NumberField but this kind of suggests the component will not be in a valid state (with the green check icon)

emircado replied:
Removed `validationState` logic from `useHexColorFieldState`. Will probably remove mentioning it in `useHexColorFieldState` once we get updates from #1029

emircado replied:
@snowystinger 
don't put validation state logic here

emircado replied:
Also just add an ignore if it's causing problems, or just leave as is

emircado replied:
Resolution: Leave as is. Currently not using it in any logic inside the hook

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,100 @@
+      ...inputProps,
+      onKeyDown: spinButtonProps.onKeyDown,
+      onFocus: spinButtonProps.onFocus,
+      onBlur: spinButtonProps.onBlur
```
I'm not sure if we need to adopt all props coming in from `spinButtonProps`. For example, if we do, `aria-maxvalue` and `aria-minvalue` would be number/decimal values of the hex strings.

<img width="470" alt="Screen Shot 2020-09-04 at 12 13 32 PM" src="https://user-images.githubusercontent.com/5743833/92277642-42272700-eea8-11ea-83ae-411e376e10da.png">

emircado replied:
Add `role `even when there's no physical buttons in the UI

yes valuenow valuetext
no valuemin valuemax

snowystinger replied:
@majornista What should the values be for say #ff88a0?
aria-valuenow: '#ff88a0' || 0xff88a0 || 16746656
aria-valuetext: '#ff88a0'??
aria-valuemin: '#000000' || 0x000000 || 0
aria-valuemax: '#FFFFFF' || 0xFFFFFF || 16777215

my understanding is that value(min/max) must be a number, so the second two options would be best?
value text though, is it just the hex string or should it be something else entirely?

majornista replied:
`aria-valuenow`, `aria-valuemin` and `aria-valuemax` must be a number:

- `aria-valuenow`: 0xff88a0 || 16746656
- `aria-valuemin`: 0x000000 || 0
- `aria-valuemax`: 0xFFFFFF || 16777215

`aria-valuetext` is a string, so for the time being it can just be the hex string that appears as the text value for the input.

If we get into trying to better express the color using localized strings for accessibility, we can overload the `aria-valuetext` property to read as something like "#ff88a0, pale pink-red". See https://www.beautycolorcode.com/ for an example of color descriptions associated with hex colors.

snowystinger replied:
thanks for the clarification
we won't worry about the color descriptions for the time being

emircado replied:
Resolved https://github.com/adobe/react-spectrum/pull/986#discussion_r493979668

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,100 @@
+    isReadOnly,
+    isRequired,
+    minValue = '#000000',
+    maxValue = '#ffffff',
```
I ended up duplicating the default min and max values here, from `useHexColorFieldState`

emircado replied:
@snowystinger: 
These are component defaults and would go to stately `useHexColorFieldState.ts`

emircado commented on packages/@react-spectrum/color/src/HexColorField.tsx:
```diff
@@ -0,0 +1,26 @@
+import React, {RefObject} from 'react';
+import {SpectrumHexColorFieldProps} from '@react-types/color';
+
+function HexColorField(props: SpectrumHexColorFieldProps, ref: RefObject<HTMLDivElement>) {
```
Resolved. Went with `RefObject<TextFieldRef>` similar to `SearchField`

emircado commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -201,3 +202,14 @@ class HSLColor implements ColorValue {
+  if (!value) return null;
+  const colorString = parseColor(value).toString('hex').substring(1);
+  return parseInt(colorString, 16);
+}
```
Check with team if this is an appropriate place for utils like these

devongovett replied:
Maybe they could be methods on `Color` itself? `parseColor` could be a `Color.parse` static method, and `parseColorToInt` could be a `color.toInt()` method?

emircado replied:
Updated! Added methods in `Color` as described.

emircado commented on packages/@react-spectrum/color/stories/HexColorField.stories.tsx:
```diff
@@ -0,0 +1,57 @@
+import React from 'react';
+import {storiesOf} from '@storybook/react';
+
+storiesOf('HexColorField', module)
```
NOTE TO SELF: Add stories for controlled and uncontrolled states

emircado commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -34,6 +35,11 @@ export class Color {
+  static parse(value: ColorInput): Color {
+    if (!value) { return null; }
+    return new Color(typeof value === 'string' ? value : value.toString('hex'));
+  }
```
NOTE TO SELF: Add test for this

emircado commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -70,6 +76,10 @@ export class Color {
 
+  toHexInt(): number {
+    return parseInt(this.toString('hex').substring(1), 16);
+  }
```
NOTE TO SELF: Add test for this

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,122 @@
+export interface HexColorFieldState extends NumberFieldState {
+  colorValue: Color,
+  setColorValue: (color: Color) => void,
+  setInputValue: (value: string) => void
```
I ended up adding `setInputValue` in the interface even when it's not in the specs #974.

It seems like the colorValue and inputValue works independently.

`colorValue`
* updates in `onChange` as the user types valid colors in hex
* shouldn't capitalize when the user inputs a valid hex color (too distracting?)

`inputValue`
* updates as the user types anything
* becomes capitalized (standard format) on blur

I feel like it violates the independence of the hook though as a react-stately package. It might not make sense why `inputValue` doesn't necessarily change when `colorValue` changes, without context from react-aria.

emircado replied:
@snowystinger 
just use setColorValue as `string` and do necessary refactor

emircado replied:
Hi @snowystinger @devongovett 

I ended up keeping `setInputValue` and having `setColorValue` only internal to `useHexColorFieldState`. 
```ts
  colorValue: Color,
  setColorValue: (color: Color) => void,  // internal only
  inputValue: string,  // already provided by NumberFieldState
  setInputValue: (value: string) => void
```

It didn't make sense to use `colorValue` and `setColorValue` for `string` types so I used `inputValue` and `setInputValue` instead. I still needed a pair like `colorValue` and `setColorValue` to handle the actual `Color` data we have.

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,121 @@
+    setColorValue,
+    inputValue,
+    setInputValue,
+    commitInputValue,
```
I feel like `inputValue` + its other methods should be in aria instead of stately. Stately should concern itself only on the underlying data structure (`Color`) and how it's represented as text in the UI/input field is up to the aria hook. 

Is this is why we call it `useHexColorFieldState` instead of `useHexColorState`? Coz it also covers states more specific to the input field like `inputValue`?

emircado replied:
@snowystinger 
Revisit this after making changes in setColorValue

emircado commented on packages/@react-stately/color/test/useHexColorFieldState.test.js:
```diff
@@ -0,0 +1,214 @@
+    expect(result.current.inputValue).toBe('#AABBCC');
+  });
+
+  it('should update colorValue (uncontrolled)', function () {
```
I'm still trying to understand the difference between controlled and uncontrolled states, especially on how it'll work considering we always allow the user to type and `colorValue` gets updated (through `onChange`) when it gets valid inputs

emircado replied:
@snowystinger 
controlled: you manage
uncontrolled: component manages its own state

emircado commented:
Hi, this is ready for a full review.

Open conversations:
1. re aria attributes https://github.com/adobe/react-spectrum/pull/986#discussion_r486684067
1. re use of `colorValue` and `setColorValue` https://github.com/adobe/react-spectrum/pull/986#discussion_r487637636

Other notes:
1. I looked through accessibility requirements for `spinbutton` and it seems `PageUp` and `PageDown` should increment/decrement by a larger amount than `ArrowUp` and `ArrowDown`. Currently the code follows only the `step` value for both key pairs.
1. A lot of tests in `HexColorField.test.js` overlaps with or duplicates tests written for `useHexColorField` and `useHexColorFieldState`. I left all the test I've written as advised.

snowystinger commented on packages/@react-spectrum/color/stories/HexColorField.stories.tsx:
```diff
@@ -0,0 +1,126 @@
+      onChange={action('change')}
+      {...props} />
+  );
+}
```
Here's a controlled story
```
  .add(
    'controlled value',
    () => <ControlledHexColorField />
  )
function ControlledHexColorField(props: any = {}) {
  let [color, setColor] = useState(new Color('#FF00AA'));
  return (
    <Flex direction="row" gap="size-100" alignItems="end">
      <HexColorField
        label="Primary Color"
        onChange={setColor}
        value={color} />
      <View width="size-400" height="size-400" UNSAFE_style={{backgroundColor: color.toString('rgb')}} />
    </Flex>
  );
}
```

snowystinger commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+import {useCallback, useState} from 'react';
+import {useControlledState} from '@react-stately/utils';
+
+export interface HexColorFieldState extends NumberFieldState {
```
```suggestion
export interface HexColorFieldState extends Omit<NumberFieldState, 'value' | 'setValue'> {
```

snowystinger replied:
this should allow you remove them from your return object and should be clearer on what the type is

emircado replied:
Resolved. Used `Omit` to remove `value` and `setValue` from the return object

snowystinger commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -34,6 +35,11 @@ export class Color {
     return x;
   }
 
+  static parse(value: ColorInput): Color {
```
Ah, I think this might be the reason that the onChange is firing so much, it keeps getting a new Color object

emircado replied:
I think `onChange` is firing frequently because I call it in `commitInputValue` whenever the HTML input's onchange is called and the string is valid.

```
  const setFieldInputValue = (value: string) => {
    setInputValue(value);
    value = value.trim();
    if (!value.length) { return; }
    if (!value.startsWith('#')) {
      value = `#${value}`;
    }
    try {
      const color = clampColor(value);
      if (color) {
        setColorValue(color);
      }
    } catch (err) {
      // ignore
    }
  };

  const commitInputValue = () => {
    setFieldInputValue(colorValue ? colorValue.toString('hex') : '');
  };
```

I previously checked how the input would behave when I had `setInputValue` in `commitInputValue` instead of `setFieldInputValue` and I wasn't getting desired behaviors in some scenarios. I can elaborate more once I start resolving this, if it still makes sense.

snowystinger replied:
sure, though just be aware that if you call setState from a useControlledState hook with the value that is already in there, it won't trigger the onChange argument, if you pass in a new object, it will, even if the new object represents the same color as the object before

emircado replied:
@snowystinger oh sorry I missed this question

> Is there a reason we need a clone of a given Color instance? I'd say lets just leave it to the constructor which takes a string and returns a new instance already, and if they already have a color, then they didn't need to parse it

I did this coz I keep getting the typescript error whenever I return just `value` if it's already a `Color` object. It's like it's not convinced the object I passed has a `value` attribute
<img width="972" alt="Screen Shot 2020-09-24 at 12 24 25 PM" src="https://user-images.githubusercontent.com/5743833/94190481-0527d200-fe61-11ea-8142-0fb7a4fcae13.png">

emircado replied:
I'll implement the `useColor` hook first and get back to this if this is still relevant

emircado replied:
Resolved by removing `Color.parse`

snowystinger commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+  const minColorInt = minColor.toHexInt();
+  const maxColorInt = maxColor.toHexInt();
+
+  const clampColor = useCallback((value: ColorInput) => {
```
lets also make it a bit more generic
clampColor(value: Color, min: Color, max: Color): Color {

now we can use this to above in useColor in order to clamp min between `#000000` and `#FFFFFE` and clamp max between `#000001` and `#FFFFFF`
(you'll need to reorder just a little)

snowystinger replied:
Also add a check in clamp that min is less than max

emircado replied:
What should happen if min is greater than max? Should they swap values or should min be decreased to be `maxColorInt - 1`?

emircado replied:
- [x] More generic `clampColor` with min, max
- [ ] `value` param in `clampColor` should be type `Color`
- [ ] check if min is less than max
- [ ] clamp min to (`#000000`, `#FFFFFE`) and max to (`#000001`, `#FFFFFF`)

emircado replied:
I'm also not sure how `value` in `clampColor` accepts only `Color` and not `ColorInput`. The props `value`, `defaultValue`, `minValue`, and `maxValue` come as `ColorInput` (string | Color) and I'm not sure at what point they all just become `Color` objects before we have the need to clamp them.

emircado replied:
Resolved. Irrelevant after spec changes removing props for custom min & max

snowystinger commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+
+  const increment = () => {
+    setColorValue((previousValue: Color) => {
+      const colorInt = previousValue ? previousValue.toHexInt() : minColorInt;
```
if this colorInt is greater than the max, just return without a new color to prevent causing too many onChange
same in the other functions below
otherwise implement one of the suggestions above so you don't create a new Color

emircado replied:
Resolved. Also updated `decrement`, `incrementToMax`, and `decrementToMin` to not call `onChange` when the value is already at min/max. Also added another check for when the input changes but the color value is the same ("#CCC <-> #CCCCCC")

snowystinger commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+    }
+  }, [minValue, minColor, setColorValue, setInputValue]);
+
+  const setFieldInputValue = (value: string) => {
```
lets ignore all characters that don't match the set `[#0-9a-fA-F]`, just trim them out before calling setInputValue, I think that'll work

emircado replied:
Alright! Does this mean the input won't accept any other keys at all that aren't in `[#0-9a-fA-F]`?

snowystinger replied:
correct, I can't think of any other keys that should be accepted, can you?

emircado replied:
Related to this change, would we add other restrictions such as typing beyond six alphanumeric characters? And allowing `#` only when it's at the beginning of the input value?

snowystinger replied:
if it's pretty easy to do, then sure

emircado replied:
If I paste a string with an invalid character in the middle "abcgcba", does it remove the invalid characters as well?

emircado replied:
- [x] Accept `[#0-9a-fA-F]` keys only
- [x] If string is pasted, remove invalid characters from entire string

snowystinger commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+      value = `#${value}`;
+    }
+    try {
+      const color = clampColor(value);
```
lets move the clamp into setColorValue, so above should be something along these lines

```
  const [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
  let safelySetColorValue = (value: Color) => {
    let newColor = clamp(value, minColor, maxColor);
    setColorValue(newColor);
  }
```
and then you'd call safelySetColorValue here and in all the other places

emircado replied:
Hi @snowystinger, I'm not sure how we can use `safelySetColorValue` to replace all `setColorValue` calls in `useHexColorFieldState`. `setColorValue` allows either a `Color` object or a function, and we pass a function in most cases so we can check for color value equality before triggering `onChange` with a new object.

emircado replied:
Resolved. Irrelevant after specs changes removing props for custom min & max

snowystinger reviewed (commented):
Hey, starting a review, I made it through most of stately today, I'll keep going tomorrow


Looks like onChange is called even when the value doesn't change
I can see this in defaultValue story by tabbing in and out repeatedly without editing the field

I also see it when i hit the max/min value and try to go past the boundary with either the scrollwheel or arrow keys

snowystinger commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+    Name            | props                                        | initExpected  | newValue  | action
+    ${'max value'}  | ${{defaultValue: '#aaa', maxValue: '#bbb'}}  | ${'#AAAAAA'}  | ${'fff'}  | ${(el) => fireEvent.change(el, {target: {value: 'fff'}})}
+    ${'min value'}  | ${{defaultValue: '#ccc', minValue: '#bbb'}}  | ${'#CCCCCC'}  | ${'000'}  | ${(el) => fireEvent.change(el, {target: {value: '000'}})}
+  `('should clamp value to $Name on change', function ({props, initExpected, newValue, action}) {
```
currently once the color is typed and before the blur happens, onChange is fired with the clamped value, we should verify that onChange call as well

emircado replied:
Resolved. Added an onChange check in this test

snowystinger reviewed (commented):
a few more comments, thanks for all the hard work you've already done, it's honestly looking really good

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,142 @@
+    value,        
+    defaultValue,
+    onChange,
+    /* eslint-disable @typescript-eslint/no-unused-vars */
```
I had the `eslint-disable` here to ignore the lines above and below
```
...
/* eslint-disable @typescript-eslint/no-unused-vars */
value,        
defaultValue,
onChange,
/* eslint-disable @typescript-eslint/no-unused-vars */
otherProps,
...
```

but your comment makes me think I'm doing it wrong 😆. I'll use the omit as recommended so these aren't necessary anymore

emircado replied:
Resolved by adding `Omit` to `props` type in `useHexColorField`

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,142 @@
+
+  return {
+    labelProps,
+    inputFieldProps: mergeProps(inputProps, {
+      role: spinButtonProps.role,
+      onKeyDown: spinButtonProps.onKeyDown,
+      onFocus: spinButtonProps.onFocus,
+      onBlur
+    })
```
Is it okay to merge the entire `spinButtonProps`? The other props such as `aria-valuenow` and `aria-valuetext` still has an open discussion here
https://github.com/adobe/react-spectrum/pull/986#discussion_r486684067

and `useSpinButton` only allows numbers to be passed into these props

snowystinger replied:
well, lets see if we can get @majornista to chime in today :)

majornista replied:
I think we should be able to merge all `spinButtonProps`.

emircado replied:
Resolved.

`useSpinButtonProps` returns `aria-disabled` and `aria-readonly` props, on top of `useTextField`'s `disabled` and `readonly`. I currently test these like below

```
  it('should return prop for disabled', function () {
    let {inputFieldProps} = renderHexColorFieldHook({isDisabled: true});
    expect(inputFieldProps['aria-disabled']).toBe(true);
    expect(inputFieldProps.disabled).toBe(true);
  });
```

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+describe('HexColorField', function () {
+  it('should handle defaults', function () {
+    renderComponent({});
+    const hexColorField = screen.getByLabelText('Primary Color');
```
I took this pattern from this article
https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen

I think the advantage of using `screen` is convenience for the most part. I'll change this so it follow the pattern the other tests have to keep it consistent instead.

snowystinger replied:
good link, thanks for sharing, maybe we'll make that move in the future, for now lets be consistent, thanks!

emircado replied:
Resolved. Uses suggested pattern now

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+    const hexColorField = screen.getByLabelText('Primary Color');
+    expect(hexColorField.value).toBe('#AABBCC');
+
+    hexColorField.focus();
```
What's the reason for using `act`? I did this initially but I realized the tests were still working fine without it

snowystinger replied:
focus didn't used to need act, however, now focus frequently results in state updates, so it needs to be wrapped in an act

emircado replied:
Resolved. Added `act` in all `focus` and `blur`

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+  it.each`
+    Name                                | expected      | action
+    ${'increment with arrow up key'}    | ${'#AAAAAE'}  | ${(el) => fireEvent.keyDown(el, {key: 'ArrowUp'})}
+    ${'increment with page up key'}     | ${'#AAAAAE'}  | ${(el) => fireEvent.keyDown(el, {key: 'PageUp'})}
```
I did notice it as well https://github.com/adobe/react-spectrum/pull/986#issuecomment-693091508
and I was hoping to get some inputs on how much to increment for page up and page down. Will leave it as is, as recommended

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+
+  it.each`
+    Name                                | expected      | action
+    ${'increment with arrow up key'}    | ${'#AAAAAE'}  | ${(el) => fireEvent.keyDown(el, {key: 'ArrowUp'})}
```
Will do! I wrote these tests following `useSpinButton.test.js` that only fired `keyDown` events

snowystinger replied:
yeah, this is something we're starting to be better about, when we first started this code base we just did the events that we needed, however, we ran into some weird cases where tests were failing or passing when they shouldn't because the events didn't match how a real user would interact

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+    expect(hexColorField).not.toBeRequired();
+  });
+
+  it('should handle label prop', function () {
```
Resolved by moving the quoted lines to the defaults test

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+    expect(hexColorField.value).toBe('#AAAAAA');
+
+    action(hexColorField);
+    expect(hexColorField.value).toBe(expected);
```
Resolved. Added an onChange check in this test and similar tests

emircado commented on packages/@react-spectrum/color/src/HexColorField.tsx:
```diff
@@ -0,0 +1,49 @@
+function HexColorField(props: SpectrumHexColorFieldProps, ref: RefObject<TextFieldRef>) {
+  props = useProviderProps(props);
+  const {
+    /* eslint-disable @typescript-eslint/no-unused-vars */
```
I was able to resolve this for `useHexColorField`, but I'm not sure how to do the same here.

```
  const {
    value,          // eslint-disable @typescript-eslint/no-unused-vars
    defaultValue,   // eslint-disable @typescript-eslint/no-unused-vars
    onChange,       // eslint-disable @typescript-eslint/no-unused-vars
    ...otherProps
  } = props;
  const state = useHexColorFieldState(props);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>();
  const {
    labelProps,
    inputFieldProps
  } = useHexColorField(otherProps, state, inputRef);
```

I pass the entire `props` for `useHexColorFieldState` but exclude some props using `otherProps` for `useHexColorField`. It doesn't seem like `Omit` removes the props you indicate on its own

snowystinger replied:
right, omit won't pull them out on their own, I see the issue you're running into, ok, lets leave it as you have it working but with the right types for now

emircado replied:
Resolved https://github.com/adobe/react-spectrum/pull/986#discussion_r493129064

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+    expect(hexColorField.value).toBe('#BBBBBB');
+  });
+
+  it('should handle uncontrolled state', function () {
```
Resolved by using `typeText` and 6-character input in this test

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+    expect(hexColorField.value).toBe('#AABBCC');
+
+    hexColorField.focus();
+    fireEvent.change(hexColorField, {target: {value: 'cba'}});
```
Resolved by using `typeText`. Also used `userEvent.clear(hexColorField)` to clear the input before `typeText` calls where needed

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,223 @@
+    expect(hexColorField.value).toBe('#AABBCC');
+
+    hexColorField.focus();
+    fireEvent.change(hexColorField, {target: {value: 'cba'}});
```
Resolved https://github.com/adobe/react-spectrum/pull/986#discussion_r493922014

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,142 @@
+    }
+  );
+
+  const onBlur = (e) => {
```
Resolved https://github.com/adobe/react-spectrum/pull/986#discussion_r493979668

emircado commented:
Ready for another review

Open discussions (I worked on these as well but had more questions):

**re updates for `clampColor`**
- https://github.com/adobe/react-spectrum/pull/986#discussion_r489824697
- https://github.com/adobe/react-spectrum/pull/986#discussion_r489816101

**re accepted characters**
- https://github.com/adobe/react-spectrum/pull/986#discussion_r489822075

I also had a general question in terms of expected behaviors. I was looking for more detailed specs ([here](https://spectrum.adobe.com/page/color-area/), [here](https://github.com/adobe/react-spectrum/issues/974)) like on how the component should handle non-hex characters (not `[#0-9a-fA-F]`) but didn't find any. I examined other color pickers I see online and they accept any character and display back the last valid color on blur. I just wanted to know how we decide on more specific behaviors

snowystinger commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,295 @@
+    expect(hexColorField.value).toBe('#AABBCC');
+
+    act(() => {hexColorField.focus();});
+    userEvent.clear(hexColorField);
```
when a field is cleared, I think we should expect onChange to be called, maybe we should have an NaN equivalent for colors or return undefined?
Otherwise if a field starts empty, a user types a color, then deletes said color, it'll never go back to being a blank empty field

emircado replied:
Progress
- [ ] wrap `userEvent` in `act`
- [ ] change `onChange` behavior for when user clears the input.

snowystinger reviewed (commented):
this is looking good
few high level issues I ran into, for some reason i can still select the text in the disabled control
This story is blowing up in storybook right now for some reason locally, are you seeing that too? story/hexcolorfield--as-a-popover-defaults-only
error message `Cannot read property 'toHexInt' of undefined`
it seems to be from a useMemo, maybe a previous color didn't exist

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,124 @@
+  ref: RefObject<HTMLInputElement>
+): HexColorFieldAria {
+  let {
+    isDisabled,
```
_(ignore the line of code cited; it's just to start this thread)_

Hi @snowystinger, regarding
> for some reason i can still select the text in the disabled control

I patterned `HexColorField.ts` from `SearchField.ts`, and I noticed you can also select text even when the input is disabled there. I'll go ahead and rework the disabled state for this component but `SearchField` behavior might be worth looking into as well

emircado replied:
<img width="566" alt="Screen Shot 2020-09-29 at 2 04 27 PM" src="https://user-images.githubusercontent.com/5743833/94615492-ba40fc80-025c-11eb-888b-a367d85a9f31.png">

emircado replied:
<img width="580" alt="Screen Shot 2020-09-29 at 2 05 10 PM" src="https://user-images.githubusercontent.com/5743833/94615573-d775cb00-025c-11eb-8c62-0daa997e96c3.png">

snowystinger replied:
oo, thanks, i only compared it to textfield, i wonder what the difference is

emircado replied:
Hi @snowystinger, I re-examined the disabled behavior for `TextField` and it turns out the text in its `isDisabled` story is not selectable because it's only a placeholder and has no value 😆. Text is selectable on a disabled `TextField`

![Screenshot 2020-10-01 164456](https://user-images.githubusercontent.com/5743833/94873946-763d2b80-0405-11eb-8e16-1d822e00765e.png)

emircado replied:
Resolving this since behavior is consistent with other components

snowystinger replied:
HAHAHAHA, awesome, thanks for figuring that out

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,150 @@
+        let newValue = `#${Math.min(prevColorInt + step, maxColorInt).toString(16).padStart(6, '0')}`;
+        newColor = new Color(newValue);
+      }
+      setInputValue(newColor.toString('hex'));
```
Will change this! My reasoning here is that I wanted to rely on `Color` to decide how the color should be represented in string, instead of using a value calculated outside of it, like line 70.

emircado replied:
Resolved for increment and decrement

emircado commented on packages/@react-spectrum/color/stories/HexColorField.stories.tsx:
```diff
@@ -0,0 +1,152 @@
+    )
+  )
+  .add(
+    'as a popover, defaults only',
```
Re
> This story is blowing up in storybook right now for some reason locally, are you seeing that too? story/hexcolorfield--as-a-popover-defaults-only
error message Cannot read property 'toHexInt' of undefined
it seems to be from a useMemo, maybe a previous color didn't exist

There was an issue with how I implemented `HexColorFieldPopover`, as it was using `useColor` provided with a value that didn't exist. Resolved now

devongovett commented:
@dannify assume you did not mean to close this...

snowystinger commented:
@emircado looks like we can simplify a little bit. Our team had a discussion about color pickers today and realized we had a miscommunication around the props. Min/Max values are not longer needed on this component. If you don't mind removing them from your PR, it'd be much appreciated.
I've updated your ticket description to remove them from the API requirements.
Thanks!

emircado commented:
@snowystinger thanks for the update! It didn't quite make sense to me as well https://github.com/adobe/react-spectrum/pull/986#issuecomment-693091508 so thanks for getting it clarified. Will update the MR with these changes as well

emircado commented:
Ready for a third review

Major changes since last review
1. Clearing the input also clears the color value selected
2. `minValue` and `maxValue` props are removed

snowystinger commented on packages/@react-spectrum/color/src/HexColorField.tsx:
```diff
@@ -0,0 +1,47 @@
+function HexColorField(props: SpectrumHexColorFieldProps, ref: RefObject<TextFieldRef>) {
+  props = useProviderProps(props);
+  let {
+    value,          // eslint-disable-line @typescript-eslint/no-unused-vars
```
I'm going to leave this one as unresolved, I still don't like the solution, but can't think of something better yet. Maybe one of my teammates will have an idea, so I want them to see this.

emircado replied:
Hi @LFDanLu, I'm not quite sure how to follow your suggestion to resolve this.

It might be fine in SearchField's case because `defaultValue` is just a string but HexColorField's `value`, `defaultValue`, and `onChange` are not string types

```ts
export interface HexColorFieldProps {
  value?: ColorInput,
  defaultValue?: ColorInput,
  onChange?: (color: Color) => void
}
```

We have to omit these props before they're used in `useHexColorField`, otherwise hooks in it like `useTextField` will throw errors because it expects these props to be `string` type only.

LFDanLu replied:
Ahh, my mistake, misunderstood the issue here. Hm will have to play around, but what you have might be the best solution.

snowystinger reviewed (commented):
Thanks for making changes so quickly!
Found an issue with the controlled state, if I delete the text, the story throws an error and doesn't recover.
One thing to keep in mind while fixing that issue is that whatever comes out of onChange must be able to be interpreted by 'value' so that controlled can work. My guess is that value can't accept null.
This is a tricky edge case, lets make sure to add a test for it, you've been amazing with all the tests you've added so far, it makes my job much easier.

snowystinger commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -70,6 +70,10 @@ export class Color {
   }
 
+  toHexInt(): number {
+    return parseInt(this.toString('hex').substring(1), 16);
```
I think we can do this a little faster, after all, hex doesn't matter, it's just a number
what if this function didn't need to parse a string and just did something like this
```
toInt(): number {
  return this.value.toRGB().toInt();
}
```

where toInt() on the RGB class is
```
return this.red * 256 + this.green * 16 + this.blue;
```

snowystinger replied:
I don't know what it'd mean for the other types, but we can start here

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+  setInputValue: (value: string) => void
+}
+
+export const minColor = new Color('#000000');
```
If I don't export them, is it okay to duplicate these constants in `useHexColorField`? There was a previous discussion on it here https://github.com/adobe/react-spectrum/pull/986#discussion_r483827650 when we went through the code before the first full review, about where to place constants like these

snowystinger replied:
lets leave it for now and see if anyone has any other opinions on it
I almost wonder if the state object should have them

emircado replied:
Progress:
- [x] rename file constants to MIN_COLOR, MAX_COLOR, MIN_COLOR_INT, MAX_COLOR_INT
- [ ] decide on where these constants should be defined, for use in `useHexColorField` and `useHexColorFieldState`

Will leave this thread open for the next review

LFDanLu replied:
I think its fine to keep these exports here. I could see it being nice to shuttle these values around in the state object if they were definable by the user though (e.g. if the HexColorField supported a notion of `minColor` and `maxColor` where the user would pass in a string and the state object would have the `Color` equivalents)

emircado replied:
Resolved https://github.com/adobe/react-spectrum/pull/986#discussion_r509721431

emircado commented on packages/@react-spectrum/color/stories/HexColorField.stories.tsx:
```diff
@@ -0,0 +1,136 @@
+        label="Primary Color"
+        onChange={onChange}
+        value={color} />
+      <View width="size-400" height="size-400" UNSAFE_style={{backgroundColor: color.toString('rgb')}} />
```
Hi @snowystinger, regarding

> Found an issue with the controlled state, if I delete the text, the story throws an error and doesn't recover.
One thing to keep in mind while fixing that issue is that whatever comes out of onChange must be able to be interpreted by 'value' so that controlled can work. My guess is that value can't accept null.
This is a tricky edge case, lets make sure to add a test for it, you've been amazing with all the tests you've added so far, it makes my job much easier.

I feel like the issue is with the story and not with the code. It doesn't check for an empty/falsey value for `color` and assumes all values of `color` will have methods like `toString('rgb')` that is throwing the error. The story would fail as well if it didn't have a `value` prop provided to it.

```
  .add(
    'controlled value',
    () => (
      <ControlledHexColorField
        value={new Color('#FF00AA')}     // when taken out, you get "Cannot read property 'toString' of undefined"
        onChange={action('change')} />
    )
  )
```

My proposed fix is to have a check first in the story before calling `Color`-specific methods.

```suggestion
function ControlledHexColorField(props: any = {}) {
  let [color, setColor] = useState(props.value);
  let onChange = (color: Color) => {
    setColor(color);
    if (props.onChange) { props.onChange(color); }
  };
  let style = color ? {backgroundColor: color.toString('rgb')} : {};
  return (
    <Flex direction="row" gap="size-100" alignItems="end">
      <HexColorField
        label="Primary Color"
        onChange={onChange}
        value={color} />
      <View width="size-400" height="size-400" UNSAFE_style={style} />
    </Flex>
  );
}
```

snowystinger replied:
o, i see, that's great news :)
then yes, I think what you're proposing is a fine fix for the story

emircado commented:
Ready for another review

Open threads
* re omitting unused props https://github.com/adobe/react-spectrum/pull/986#discussion_r496268067
* re where to place constants https://github.com/adobe/react-spectrum/pull/986#discussion_r498604978

snowystinger commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -123,6 +128,10 @@ class RGBColor implements ColorValue {
   }
 
+  toInt() {
+    return this.red * Math.pow(16, 4) + this.green * Math.pow(16, 2) + this.blue;
```
Math.pow(16, 4) is a constant, same with 16, 2 lets only calculate them once when the module loads

snowystinger reviewed (dismissed):
Looks good, we'll get some other reviews to try and resolve those few remaining open items

majornista commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+    inputValue,
+    setInputValue: setFieldInputValue,
+    commitInputValue,
+    increment,
+    incrementToMax,
+    decrement,
+    decrementToMin,
```
Increment and Decrement are not particularly useful one integer value at a time, I wonder if `incrementPage` and `decrementPage` could be added to increment and decrement, snapping to the closest Web Safe or named color value.

snowystinger replied:
Don't worry about this one in this PR

majornista reviewed (commented):
Increment and Decrement are not particularly useful one integer value at a time, I wonder if incrementPage and decrementPage could be added at [R127-R130](https://github.com/adobe/react-spectrum/pull/986/files#diff-2e6fbd95c56ba13b819b8460f0144bf4R127-R130) to increment and decrement, snapping to the closest Web Safe or named color value.

LFDanLu reviewed (changes requested):
Looks good to me for the most part. The scroll wheel firing onChange twice is odd, but the fix maybe in a similar vein as https://github.com/adobe/react-spectrum/issues/1158. I see something similar happening in NumberField so I'd be fine with ignoring it for this pull

emircado commented on packages/@react-spectrum/color/test/HexColorField.test.js:
```diff
@@ -0,0 +1,299 @@
+    expect(hexColorField.value).toBe('#AAAAAA');
+
+    fireEvent.wheel(hexColorField, {deltaY});
+    expect(onChangeSpy).toHaveBeenCalledWith(expected);
```
Somehow I was able to add an expect for `onChangeSpy` to be called only once and the test passed.

So far, I noticed the mousewheel event handler in `useHexColorField` only calls increment/decrement once but the function passed to `setColorValue` in `useHexColorFieldState` gets called twice.

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,118 @@
+
+      e.preventDefault();
+      if (e.deltaY < 0) {
+        increment();
```
So far, I noticed the mousewheel event handler in `useHexColorField` only calls increment/decrement once but the function passed to `setColorValue` in `useHexColorFieldState` gets called twice. Still looking into it.

emircado replied:
Hi @LFDanLu @snowystinger, I managed to have the wheel event call `setColorValue` once by returning an `onWheel` prop from `useHexColorField`, instead of adding event listeners

suggestion
```ts
// this is returned in useHexColorField inside inputProps
  let onWheel = (e) => {
    if (isDisabled || isReadOnly) {
      return;
    }
    if (e.deltaY < 0) {
      increment();
    } else {
      decrement();
    }
  };
```

current code
```ts
  useEffect(() => {
    let currentRef = ref.current;
    let handleInputScrollWheel = e => {
      // If the input isn't supposed to receive input, do nothing.
      // TODO: add focus
      if (isDisabled || isReadOnly || !currentRef) {
        return;
      }

      e.preventDefault();
      if (e.deltaY < 0) {
        increment();
      } else {
        decrement();
      }
    };

    if (!currentRef) { return; }
    currentRef.addEventListener(
      'wheel',
      handleInputScrollWheel,
      {passive: false}
    );
    return () => {
      currentRef.removeEventListener(
        'wheel',
        handleInputScrollWheel
      );
    };
  }, [inputId, isReadOnly, isDisabled, decrement, increment, ref]);
```

I'm not sure what the side-effect of this approach would be. Please advise

snowystinger replied:
o, cool, yeah, not sure why we aren't doing that already... that seems more right for this

LFDanLu replied:
+1 to ^, I think returning `onWheel` as a part of inputProps makes sense

emircado commented:
Still open for reviews

Open discussions
1. re how to omit unused props - https://github.com/adobe/react-spectrum/pull/986#discussion_r504993013
2. re where to put constants - https://github.com/adobe/react-spectrum/pull/986#discussion_r499066695

devongovett commented on packages/@react-stately/color/src/Color.ts:
```diff
@@ -162,6 +169,9 @@ class RGBColor implements ColorValue {
   }
 
+  toInt() {
+    return this.red * RED_MULTIPLIER + this.green * GREEEN_MULTIPLIER + this.blue;
```
You could use bit ops for this: 😄 

```suggestion
    return this.red << 16 | this.green << 8 | this.blue;
```

This works because each component is 8 bits (0-255), so you shift each component over by 8 bits and combine them together with bitwise or so they sit next to each other in a single bitstring.

snowystinger replied:
nice! might be clearer to read, but may not be more efficient i don't think
to do the bit operators all the numbers have to be converted from 64 bit floating to 32 bit signed first
I don't think anything will be happening fast enough to worry about the speed though

devongovett reviewed (commented):
This looks great! My comments are mostly suggestions for improvements. I'm happy to merge this as an initial version and treat these as followups too. 😄

emircado commented on packages/@react-aria/color/src/useHexColorField.ts:
```diff
@@ -0,0 +1,96 @@
+      id: inputId,
+      value: inputValue,
+      type: 'text',
+      autoComplete: 'off',
```
`useTextField` didn't seem to recognize `autoCorrect` when I passed it, so I just added it directly to the return object

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+  let [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
+
+  let initialInputValue = (value || defaultValue) && colorValue ? colorValue.toString('hex') : '';
+  let [inputValue, setInputValue] = useState(initialInputValue);
```
Hi @devongovett, I'm not sure I understand your comment 😅.

In cases where changes in the input value lead the `value` change (like typing), the input doesn't update (`commitInputValue`) until you blur out of the field, so the input doesn't disrupt you and change the input text to uppercase when you're typing in lowercase.

In cases where you don't type (increment, decrement via keyboard/mouse), the input value already updates along with the change in color value

devongovett replied:
I mean when the `value` prop updates from the outside:

```jsx
// Before
<HexColorField value="#abcdef" />

// After
<HexColorField value="#123456" />
```

I believe right now the input value would still stay as `#abcdef` after the prop change because the `inputValue` state is not updated.

snowystinger replied:
Here's how I handled it in numberfield, i ran into this one as well https://github.com/adobe/react-spectrum/pull/1029/files#diff-f920ff47503335092fb8b412b9e4d0679256fecedc505827ad70a7c7871354dbR104
and here's the test i wrote for it
https://github.com/adobe/react-spectrum/pull/1029/files#diff-89757429f0981d25f225267128d9551d5dee758c05de1c49d2ae1dd09fcb7306R1063

emircado replied:
Hi @devongovett @snowystinger,

I'm now able to have the input updated when a new `value` prop is passed by adding the snippet below in stately
```ts
  useEffect(() => {
    setInputValue(colorValue ? colorValue.toString('hex') : '');
  }, [colorValue, setInputValue]);
```

But now it's updating the input unnecessarily (I think). It reformats the string as you delete characters coz they are still valid color strings.
![d](https://user-images.githubusercontent.com/5743833/97055778-5a2c3600-153c-11eb-8698-3083d32066a9.gif)

devongovett replied:
Ah good catch! Maybe you could only update the inputValue if the color coming from props is different from the one that's already in state. Something like this (untested):

```jsx
useEffect(() => {
  setInputValue(inputValue => {
    // Parse color from current inputValue.
    // Only update the input value if the new colorValue is not equivalent.
    let currentColor = new Color(inputValue);
    if (currentColor?.toHexInt() !== colorValue?.toHexInt()) {
      return colorValue ? colorValue.toString('hex') : '';
    }

    return inputValue;
  });
}, [colorValue, setInputValue]);
```

emircado replied:
Hi @devongovett, I've committed changes to resolve this issue, but I'm having issues writing the test for it.

I wrote the test below
```js
  it('should update value in controlled state', function () {
    function HexColorFieldControlled(props) {
      let {onChange} = props;
      let [color, setColor] = useState(props.value);
      return (
        <HexColorField
          {...props}
          label="Primary Color"
          value={color}
          onChange={chain(setColor, onChange)} />
      ); 
    }
    let onChangeSpy = jest.fn();
    let {getByLabelText} = render(<HexColorFieldControlled value={new Color('#abc')} onChange={onChangeSpy} />);

    let hexColorField = getByLabelText('Primary Color');
    expect(hexColorField.value).toBe('#AABBCC');

    act(() => {
      hexColorField.focus();
      userEvent.clear(hexColorField);
    });
    // should call onChange when input is cleared
    expect(hexColorField.value).toBe('');
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
    expect(onChangeSpy).toHaveBeenCalledWith(null);

    typeText(hexColorField, 'cba');
    expect(hexColorField.value).toBe('cba');
    expect(onChangeSpy).toHaveBeenCalledTimes(2);
    expect(onChangeSpy).toHaveBeenCalledWith(new Color('#cba'));

    act(() => {hexColorField.blur();});
    expect(hexColorField.value).toBe('#CCBBAA');
    expect(onChangeSpy).toHaveBeenCalledTimes(2);
  });
```

but I'm getting the warning below
![Screenshot 2020-10-26 173917](https://user-images.githubusercontent.com/5743833/97242873-34a86200-17b2-11eb-9409-622532780d31.png)

I also see it in the stories I have where I have controlled components, after triggering onChange events. I've looked it up in multiple places but haven't got closer to fixing 😆. It's something about calling setState functions while the component is rendering

emircado replied:
resolved warning above with https://github.com/adobe/react-spectrum/pull/986/commits/813dcf592aba39e34179d39edc260c261c7764c2

emircado commented on packages/@react-stately/color/src/useHexColorFieldState.ts:
```diff
@@ -0,0 +1,133 @@
+  }, [setColorValue, setInputValue]);
+
+  let setFieldInputValue = (value: string) => {
+    value = value.replace(/[^#0-9a-f]/ig, '');
```
I updated the regex to `{0,6}` from `{1,6}` because I was having issues. It wasn't letting me delete (backspace) when there is one character left coz it doesn't match the pattern.

emircado commented:
Ready for another review

devongovett reviewed (dismissed):
Looks great! We'll merge after the release tomorrow morning. 🥳🎉
